File: LeetCode-C++/LeetCode-C++/二叉树/_104_二叉树的最大深度.cpp
//
//  _104_二叉树的最大深度.cpp
//  LeetCode-C++
//
//  Created by warden on 2019/12/11.
//  Copyright © 2019 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/maximum-depth-of-binary-tree/

#include <stdio.h>
#include <queue>
#include "TreeNode.hpp"

using namespace std;

class Solution {
public:
    int maxDepth(TreeNode* root) {
        int depth = 0;
        queue<TreeNode*> nodes;
        if (root) {
            nodes.push(root);
        }
        size_t size = 0;
        while (!nodes.empty()) {
            if (size == 0) {
                size = nodes.size();
                depth++;
            }
            TreeNode* p = nodes.front();
            nodes.pop();
            if (p->left) {
                nodes.push(p->left);
            }
            if (p->right) {
                nodes.push(p->right);
            }
            size--;
        }
        return depth;
    }
};

/**
 示例：
 给定二叉树 [3,9,20,null,null,15,7]，

     3
    / \
   9  20
     /  \
    15   7
 返回它的最大深度 3 。
 */
/**
int main(int argc, const char * argv[]) {
    TreeNode* root = generateTreeByLevel({3, 9, 20, INT_MAX, INT_MAX, 15, 7});
    Solution* solution = new Solution();
    printf("%d\n", solution->maxDepth(root));
    return 0;
}
 */

File: LeetCode-Swift/LeetCode-Swift/二叉树/_104_二叉树的最大深度.swift
//
//  _104_二叉树的最大深度.swift
//  LeetCode-Swift
//
//  Created by warden on 2019/12/11.
//  Copyright © 2019 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/maximum-depth-of-binary-tree/

import Foundation

class _104_二叉树的最大深度 {
    func maxDepth(_ root: TreeNode?) -> Int {
        if root == nil {
            return 0
        }
        var depth = 0
        var nodes = [root!]
        var size = 0
        while !nodes.isEmpty {
            if size == 0 {
                size = nodes.count
                depth += 1
            }
            let p = nodes.removeFirst()
            if p.left != nil {
                nodes.append(p.left!)
            }
            if p.right != nil {
                nodes.append(p.right!)
            }
            size -= 1
        }
        return depth
    }
}

/**
 给定二叉树 [3,9,20,null,null,15,7]，

     3
    / \
   9  20
     /  \
    15   7
 返回它的最大深度 3
 */
func test_104() {
    let root = generateTreeByLevel([3, 9, 20, nil, nil, 15, 7])
//    let root: TreeNode? = nil
    let solution = _104_二叉树的最大深度()
    print(solution.maxDepth(root))
}
